### PR TITLE
Sequence rewards

### PR DESCRIPTION
**Pull request by unconst**
Build multi layer defences by stacking reward functions.

### CODE REVIEW

**Comment by mrseeker** [comment]
Would recommend having the blacklist be a regex test, rather than a single sentence. This way it's much harder to exploit. "That's an excellent question!" is not blocked, the "That is an excellent question." is despite being both being the same annoying message.

**Review by opentaco — APPROVED** [review]
Definitely an enabling reorg, would also allow for constitutional prompt-based reward models to be added in a similar manner.

May need to switch majority of reward models off by default, since the default all-enabled could exceed VRAM with so many models.

**Comment by mrseeker** [comment]
Another thought is to have the "diversity" reward model work differently, and use a "cosine similarity" test:

https://towardsdatascience.com/how-to-filter-out-similar-texts-in-python-c7e7c5f7620e

It's easier to detect "cached" results, even when they are differently worded (to avoid the "diversity" filter). The score should reflect the "uniqueness" of the answer compared to all other answers. The more unique the answer is, the better the score. This should be combined with something like a relevance filter and should not be used "stand-alone".

File: openvalidators/reward/diversity.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    def reward( self, prompt: str, completion: str ) -> float:
+        if prompt not in self.counters: 
+            self.counters[ prompt ] = {}
+        
+        if completion not in self.counters[ prompt ]:
+            self.counters[ prompt ][ completion ] = 0
+
+        self.counters[ prompt ][ completion ] += 1
+        reward = 1.0 / self.counters[ prompt ][ completion ]
```
Comment (adriansmares) [inline openvalidators/reward/diversity.py:37]: `counters` needs to be reset between batches, and in general it really needs the full completions batch in order to operate. Otherwise the division depends on the order of evaluation, which is random.

Comment (unconst) [reply openvalidators/reward/diversity.py:37]: noted. Mostly a placeholder rn.

File: openvalidators/reward/relevance.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            embeddings = self.model(**encoded_input)
+
+        sentence_embeddings = BertRelevanceRewardModel.mean_pooling(embeddings, encoded_input["attention_mask"])
+        sentence_embeddings = torch.nn.functional.normalize(sentence_embeddings, p=2, dim=1)
+        batch_representation = torch.mean(sentence_embeddings, dim=0)
+        return batch_representation
+    
+
+    def reward( self, prompt: str, completion: str ) -> float:
+
```
[the remark this answers is not in the document]
Comment (unconst) [reply openvalidators/reward/relevance.py:87]: change it yours @steffencruz

File: openvalidators/reward/diversity.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (unconst) [reply openvalidators/reward/diversity.py]: danke

File: openvalidators/reward/dahoas.py [hunk 3]
```diff
@@ -15,22 +15,40 @@
...
 
-class RewardModel(torch.nn.Module):
-    def __init__(self, model_path: str, device: str, config: "bt.config" = None):
+    model_name = "EleutherAI/gpt-j-6b"
+
+    @property
+    def name(self) -> str: return "dahoas-reward-model"
+
+    @staticmethod
+    def load_weights( path: str ):
```
[the remark this answers is not in the document]
Comment (unconst) [reply openvalidators/reward/dahoas.py:34]: done

File: openvalidators/neuron.py [hunk 4]
```diff
@@ -114,6 +123,23 @@ def __init__(self):
             self.dendrite_pool = AsyncDendritePool(keypair=self.wallet.hotkey, metagraph=self.metagraph)
         bt.logging.debug(str(self.dendrite_pool))
 
+        # Init Reward model
+        bt.logging.debug("loading", "reward_functions")
+        if self.config.neuron.mock_reward_models:
+            self.reward_functions = [ MockRewardModel('mock-blacklist'), MockRewardModel('mock-nsfw') ]
```
[the remark this answers is not in the document]
Comment (unconst) [reply openvalidators/neuron.py:130]: this helps us determine if the item is real or mock, why remove?

File: openvalidators/config.py [hunk 5]
```diff
@@ -217,48 +195,77 @@ def add_args(cls, parser):
...
         action="store_true",
         help="Dont download the reward model.",
         default=False,
     )
     parser.add_argument(
-        "--neuron.mock_dendrite_pool",
+        "--neuron.blacklist_off",
         action="store_true",
-        help="Dont download the dendrite pool.",
+        help="Dont apply the nsfw reward model",
```
[the remark this answers is not in the document]
Comment (unconst) [reply openvalidators/config.py:206]: resolved.

File: openvalidators/config.py [hunk 6]
```diff
@@ -217,48 +195,77 @@ def add_args(cls, parser):
...
+    parser.add_argument(
+        "--neuron.dahoas_off",
+        action="store_true",
+        help="Dont apply the dahoas reward model",
+        default=False,
+    )
+    parser.add_argument(
+        "--neuron.mock_dendrite_pool",
+        action="store_true",
+        help="Dont download the dendrite pool.",
```
[the remark this answers is not in the document]
Comment (unconst) [reply openvalidators/config.py:254]: resolved.

File: openvalidators/neuron.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (steffencruz) [reply openvalidators/neuron.py:130]: the class name contains mock, so the arg doesnt need to repeat it

File: openvalidators/prompts.py [hunk 7]
```diff
@@ -119,6 +119,16 @@ def find_unique_tags(input_text: str):
...
 followup_request_template = "Ask one relevant and insightful question about the preceding context."
 
+# ==============================
+# === Augmentation request prompt ===
+# ==============================
+
+# Request an augmentation given a preceding context.
+
+augment_request_template = "Summarize the preceding context at"
+school_levels = ['elementary school','middle school', 'highschool', 'university', 'graduate school']
```
Comment (shibshib) [inline openvalidators/prompts.py]: As per the conversation with @Eugene-hu , I think these school_levels are too basic and too close together (will a response truly be that different between elementary and middle school?) 

We should add more randomization here. "Explain why this summarization is correct or incorrect", "Summarize the preceding text in pig latin", "Summarize the preceding context at a level that my grandmother can understand" etc.

File: openvalidators/reward/reciprocate.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    def __init__( self, device: str ):
+        super().__init__()
+        self.device = device
+        self.tokenizer = AutoTokenizer.from_pretrained( ReciprocateRewardModel.reward_model_path, revision = ReciprocateRewardModel.revision )
+        self.model = AutoModelForSequenceClassification.from_pretrained( ReciprocateRewardModel.reward_model_path, revision = ReciprocateRewardModel.revision).to(self.device)
+
+    def reward( self, prompt: str, completion: str, name: str ) -> float:
+        with torch.no_grad():
+            message = f"<|prompter|>{prompt}</s><|assistant|>{completion}</s><|endoftext|>"
```
Comment (adriansmares) [inline openvalidators/reward/reciprocate.py:41]: `completion` should be sanitized for tokens that could affect the model (what if completion contains `</s>` or `<|endoftext|>`).

**Review by adriansmares — COMMENTED** [review]
How much VRAM do validators require now ? This should be announced before they auto update, otherwise they will just restart loop indefinitely on smaller cards.

File: openvalidators/reward/reciprocate.py (hunk 8 above)
Comment (shibshib) [reply openvalidators/reward/reciprocate.py:41]: Agree here. It may make sense to just run `completion` through a regex, just in case. 

``` 
import re
regex_cleaner = re.compile('<.*?>') 
completion = re.sub(regex_cleaner, '', completion)
```

**Review by shibshib — APPROVED** [review]
